# 5. 변동성  돌파 전략 구현 및 ETF 자동 매매 시스템 구축 

#### [개요] <br>
래리 윌리엄스의 변동성 돌파 전략과 대신증권 크레온 API을 이용해 자동 매매 시스템을 구축해보고자 한다. <br>
매일 아침 자동매매 프로그램이 구동되어 대신증권에 로그인한 후 변동성 돌파 전략에 따라 자동으로 매수/매도 주문을 내도록 크레온 API를 이용한다. <br>
슬랙 API를 이용하여 주문 및 처리 결과 메시지를 실시간으로 스마트폰으로 전달하는 기능까지 구현하는 것을 목표로 한다. 

## 5.1 래리 윌리엄스의 변동성 돌파 전략 

### 변동성 돌파 전략
    가격이 전일 가격 범위의  K% 이상이 될 때 매수한 후 장마감 시 매도해서 수익을 실현하는 단기 트레이닝 기법 
    (추세 한 번 형성 시 가격은 계속 그 방향으로 움직인다는 추세 추종 이론 기반)
                     
    - 일 단위로 이루어지는 단기 매매기법 ( 매매 기간이 하루를 넘기기 않고 무조건 종가에 매도하기 때문에 장마감 후 악재가 발생하더라도 매매에 영향 미치지 않음. 
                                         다음 날 시가에 매도하는 것보다 상대적으로 더 안전.)    
    [매매 단계] 
        1. 전날의 고가에서 저가를 빼서 가격 변동폭 계산
        2. 장중 가격이 오늘 시가 + 전일 변동폭 * K를 돌파 시 매수 (일반적으로 K값 0.5, 수익률에 따라 조정 필요)
        3. 장마감 시 무조건 매도 
        
    [변동성 돌파 전략 예시]
           - 전일 시가: $920, 전일 고가 : $1000, 전일 저가 : $900, 전일 종가 $960
           - 오늘 시가: $960  
       - 돌파 가격 = $960 + ($1000 - $900) * 0.6 = $1020
       - 오늘 장중에 돌파 가격 $1020을 넘어서면 매수해서 다음 날 아침 장 시작하면 모두 매도

## 5.2 크레온 플러스 API 

    Python으로 사용 가능한 증권 API는 키움증권, 대신증권, 이베스트증권이 있다. 
    그 중, 함수 호출 즉시 결괏값을 받을 수 있어 직관적인 프로그래밍이 가능한 대신증권의 '크레온 플러스 API'를 사용하고자 한다. 

### 5.2.1 크레온 로그인 
    - 크레온 계좌 생성 
    - 크레온 플러스 api 설치 

### 5.2.2 크레온 플러스 자동 접속
    - 관리자 권한 실행 적용
    - 크레온 플러스 접속 과정 자동화 라이브러리 설치 및 코드 실행 
    

In [1]:
import os, sys, ctypes
import win32com.client
import pandas as pd
from datetime import datetime
from slacker import Slacker
import time, calendar
from bs4 import BeautifulSoup
from urllib.request import urlopen
from selenium import webdriver
from selenium.webdriver.chrome.options import Options 
import platform
print(platform.architecture())
from pywinauto import application
import time
import os

('32bit', 'WindowsPE')


In [2]:
os.system('taskkill /IM coStarter* /F /T')
os.system('taskkill /IM CpStart* /F /T')
os.system('taskkill /IM DibServer* /F /T')
os.system('wmic process where "name like \'%coStarter%\'" call terminate')
os.system('wmic process where "name like \'%CpStart%\'" call terminate')
os.system('wmic process where "name like \'%DibServer%\'" call terminate')
time.sleep(5)        

app = application.Application()
app.start('C:\CREON\STARTER\coStarter.exe /prj:cp /id:minchan* /pwd:min2468! /pwdcert:minchan4108! /autostart')
time.sleep(60)

In [12]:
#slack = Slacker('xoxb-3754340420997-3742673441015-kiC0wYN6j11UxOqAxWr13MXS')
def dbgout(message):
    """인자로 받은 문자열을 파이썬 셸과 슬랙으로 동시에 출력한다."""
    print(datetime.now().strftime('[%m/%d %H:%M:%S]'), message)
    strbuf = datetime.now().strftime('[%m/%d %H:%M:%S] ') + message
    #slack.chat.post_message('#etf-algo-trading', strbuf)

def printlog(message, *args):
    """인자로 받은 문자열을 파이썬 셸에 출력한다."""
    print(datetime.now().strftime('[%m/%d %H:%M:%S]'), message, *args)
    
 # 크레온 플러스 공통 OBJECT
cpCodeMgr = win32com.client.Dispatch('CpUtil.CpStockCode')
cpStatus = win32com.client.Dispatch('CpUtil.CpCybos')
cpTradeUtil = win32com.client.Dispatch('CpTrade.CpTdUtil')
cpStock = win32com.client.Dispatch('DsCbo1.StockMst')
cpOhlc = win32com.client.Dispatch('CpSysDib.StockChart')
cpBalance = win32com.client.Dispatch('CpTrade.CpTd6033')
cpCash = win32com.client.Dispatch('CpTrade.CpTdNew5331A')
cpOrder = win32com.client.Dispatch('CpTrade.CpTd0311')  

### 5.2.3 크레온 시스템 접속 체크 

In [3]:
cpStatus = win32com.client.Dispatch('CpUtil.CpCybos')
cpTradUtil = win32com.client.Dispatch('CpTrade.CpTdUtil') 

def check_creon_system():
    """크레온 플러스 시스템 연결 상태를 점검한다."""
    # 관리자 권한으로 프로세스 실행 여부
    if not ctypes.windll.shell32.IsUserAnAdmin():
        printlog('check_creon_system() : admin user -> FAILED')
        return False
 
    # 연결 여부 체크
    if (cpStatus.IsConnect == 0):
        printlog('check_creon_system() : connect to server -> FAILED')
        return False
 
    # 주문 관련 초기화 - 계좌 관련 코드가 있을 때만 사용
    if (cpTradeUtil.TradeInit(0) != 0):
        printlog('check_creon_system() : init trade -> FAILED')
        return False
    return print('크레온 API 정상 접속 및 연동 확인되었습니다.')

In [4]:
check_creon_system()

크레온 API 정상 접속 및 연동 확인되었습니다.


## 5.3 주가 및 계좌 정보 조회 

### 5.3.1 현재가 조회

In [5]:
def get_current_price(code):
    """인자로 받은 종목의 현재가, 매도호가, 매수호가를 반환한다."""
    cpStock.SetInputValue(0, code)  # 종목코드에 대한 가격 정보
    cpStock.BlockRequest()
    item = {}
    item['cur_price'] = cpStock.GetHeaderValue(11)   # 현재가
    item['ask'] =  cpStock.GetHeaderValue(16)        # 매도호가
    item['bid'] =  cpStock.GetHeaderValue(17)        # 매수호가    
    
    return item['cur_price'], item['ask'], item['bid']

In [6]:
# TIGER 미국채10년선물의 현재가 조회 
print(get_current_price('A305080')) 
print('현재가, 매도호가, 매수호가')

(12050, 12050, 12045)
현재가, 매도호가, 매수호가


### 5.3.2 OHLC 조회
    크레온 API로 일별 시세 조회하는 함수 
    종목코드와 수량을 입력받아서 수량에 해당하는 기간 동안의 일별 OHLC 시세를 표시 

In [7]:
cpOhlc = win32com.client.Dispatch('CpSysDib.StockChart') # OHLC 정보

def get_ohlc(code, qty):
    """인자로 받은 종목의 OHLC 가격 정보를 qty 개수만큼 반환한다."""
    cpOhlc.SetInputValue(0, code)           # 종목코드
    cpOhlc.SetInputValue(1, ord('2'))        # 1:기간, 2:개수
    cpOhlc.SetInputValue(4, qty)             # 요청개수
    cpOhlc.SetInputValue(5, [0, 2, 3, 4, 5]) # 0:날짜, 2~5:OHLC
    cpOhlc.SetInputValue(6, ord('D'))        # D:일단위
    cpOhlc.SetInputValue(9, ord('1'))        # 0:무수정주가, 1:수정주가
    cpOhlc.BlockRequest()
    count = cpOhlc.GetHeaderValue(3)   # 3:수신개수
    columns = ['open', 'high', 'low', 'close']
    index = []
    rows = []
    for i in range(count): 
        index.append(cpOhlc.GetDataValue(0, i)) 
        rows.append([cpOhlc.GetDataValue(1, i), cpOhlc.GetDataValue(2, i),
            cpOhlc.GetDataValue(3, i), cpOhlc.GetDataValue(4, i)]) 
    df = pd.DataFrame(rows, columns=columns, index=index) 
    return df

In [8]:
get_ohlc('A305080', 15)

open   high    low  close
20220705  12055  12055  11980  12050
20220704  12060  12105  12045  12055
20220701  11900  12035  11870  12000
20220630  12000  12000  11870  11900
20220629  11700  11855  11700  11855
20220628  11610  11725  11610  11680
20220627  11890  11890  11685  11730
20220624  11895  11920  11850  11890
20220623  11835  11890  11830  11850
20220622  11685  11740  11675  11740
20220621  11675  11685  11610  11685
20220620  11650  11790  11650  11675
20220617  11615  11700  11525  11650
20220616  11555  11615  11515  11605
20220615  11630  11630  11500  11555

### 5.3.3 주식 잔고 조회
    크레온 계좌에 어떤 주식 종목이 얼마의 수량만큼 있는지 조회하는 함수
    인수로 특정 주식의 종목코드를 받으면 해당 종목의 종목명과 수량을 반환하고, 
    인수로 'ALL'을 받으면 현재 계좌에 보유 중인 모든 종목에 대해 종목명과 수량을 반환

In [9]:
# cpTradeUtil = win32com.client.Dispatch('CpTrade.CpTdUtil')
# cpBalance = win32com.client.Dispatch('CpTrade.CpTd6033')
# cpCodeMgr = win32com.client.Dispatch('CpUtil.CpStockCode')

def get_stock_balance(code):
    """인자로 받은 종목의 종목명과 수량을 반환한다."""
    cpTradeUtil.TradeInit()
    acc = cpTradeUtil.AccountNumber[0]      # 계좌번호
    accFlag = cpTradeUtil.GoodsList(acc, 1) # -1:전체, 1:주식, 2:선물/옵션
    cpBalance.SetInputValue(0, acc)         # 계좌번호
    cpBalance.SetInputValue(1, accFlag[0])  # 상품구분 - 주식 상품 중 첫번째
    cpBalance.SetInputValue(2, 50)          # 요청 건수(최대 50)
    cpBalance.BlockRequest()     
    if code == 'ALL':
        dbgout('계좌명: ' + str(cpBalance.GetHeaderValue(0)))
        dbgout('결제잔고수량 : ' + str(cpBalance.GetHeaderValue(1)))
        dbgout('평가금액: ' + str(cpBalance.GetHeaderValue(3)))
        dbgout('평가손익: ' + str(cpBalance.GetHeaderValue(4)))
        dbgout('종목수: ' + str(cpBalance.GetHeaderValue(7)))
    stocks = []
    for i in range(cpBalance.GetHeaderValue(7)):
        stock_code = cpBalance.GetDataValue(12, i)  # 종목코드
        stock_name = cpBalance.GetDataValue(0, i)   # 종목명
        stock_qty = cpBalance.GetDataValue(15, i)   # 수량
        if code == 'ALL':
            dbgout(str(i+1) + ' ' + stock_code + '(' + stock_name + ')' 
                + ':' + str(stock_qty))
            stocks.append({'code': stock_code, 'name': stock_name, 
                'qty': stock_qty})
        if stock_code == code:  
            return stock_name, stock_qty
    if code == 'ALL':
        return stocks
    else:
        stock_name = cpCodeMgr.CodeToName(code)
        return stock_name, 0

In [10]:
# 현재 보유하고 있는 'TIGER 미국채10년선물' 종목의 개수이다.
get_stock_balance("A305080")

('TIGER 미국채10년선물', 0)

In [13]:
# 현재 내 계좌에 보유 중인 종목과 수량
get_stock_balance('ALL')

[07/05 13:49:30] 계좌명: 변민찬
[07/05 13:49:30] 결제잔고수량 : 0
[07/05 13:49:30] 평가금액: 0
[07/05 13:49:30] 평가손익: 0
[07/05 13:49:30] 종목수: 0


[]

### 5.3.4 주문 가능 금액 조회

In [16]:
cpTradeUtil = win32com.client.Dispatch('CpTrade.CpTdUtil')
cpCash = win32com.client.Dispatch('CpTrade.CpTdNew5331A')

def get_current_cash():
    """증거금 100% 주문 가능 금액을 반환한다."""
    cpTradeUtil.TradeInit()
    acc = cpTradeUtil.AccountNumber[0]    # 계좌번호
    accFlag = cpTradeUtil.GoodsList(acc, 1) # -1:전체, 1:주식, 2:선물/옵션
    cpCash.SetInputValue(0, acc)              # 계좌번호
    cpCash.SetInputValue(1, accFlag[0])      # 상품구분 - 주식 상품 중 첫번째
    cpCash.BlockRequest() 
    return cpCash.GetHeaderValue(9) # 증거금 100% 주문 가능 금액

In [17]:
print(get_current_cash())
print('0 이 맞다. 계좌에 금액이 없기 때문이다.')

0
0 이 맞다. 계좌에 금액이 없기 때문이다.


## 5.4 ETF 매수/매도
    변동성 돌파 전략은 일반적인 주식 종목뿐만 아니라 선물, ETF, 암호화폐 등 다양한 금융 자산에 적용할 수 있다.
    
    ETF: 상장지수펀드, 인덱스 펀드를 거래소에 상장시켜 주식처럼 편리가헤 매매할 수 있는 상품
    그 중, ETF를 선택한 이유는 주식처럼 거래마다 세금을 내지 않아도 되는 것! 

### 5.4.1 네이버 ETF 정보 스크래핑

In [18]:
from selenium import webdriver
from bs4 import BeautifulSoup
import pandas as pd

# 옵션값 설정
opt = webdriver.ChromeOptions()
opt.add_argument('headless')

# 웹드라이버를 통해 네이버 금융 ETF 페이지에 접속
drv = webdriver.Chrome('C:\myPackage\chromedriver.exe', options=opt)
drv.implicitly_wait(3)
drv.get('https://finance.naver.com/sise/etf.nhn')

# 뷰티풀 수프로 테이블을 스크래핑
bs = BeautifulSoup(drv.page_source, 'lxml')
drv.quit()
table = bs.find_all("table", class_="type_1 type_etf")
df = pd.read_html(str(table), header=0)[0]

# 불필요한 열과 행을 삭제하고 인덱스를 재설정해서 출력
df = df.drop(columns=['Unnamed: 9'])
df = df.dropna()
df.index = range(1, len(df)+1)
print(df)

# 링크 주소에 포함된 종목코드를 추출하여 전체 종목코드와 종목명 출력
etf_td = bs.find_all("td", class_="ctg")
etfs = {}
for td in etf_td:
    s = str(td.a["href"]).split('=')
    code = s[-1]
    etfs[td.a.text] = code
print("etfs :", etfs)

df = df.sort_values(by=['거래대금(백만)'], ascending=False)
df

C:\Users\mnch\AppData\Local\Temp\ipykernel_21748\3077744575.py:10: DeprecationWarning: executable_path has been deprecated, please pass in a Service object
  drv = webdriver.Chrome('C:\myPackage\chromedriver.exe', options=opt)


                       종목명      현재가    전일비     등락률      NAV   3개월수익률  \
1                KODEX 200  30930.0  360.0  +1.18%  30960.0  -16.12%   
2    TIGER 차이나전기차SOLACTIVE  17840.0   10.0  +0.06%  17885.0  +21.17%   
3               KODEX 레버리지  14360.0  345.0  +2.46%  14394.0  -30.98%   
4           TIGER 미국나스닥100  67700.0  715.0  +1.07%  67427.0  -17.18%   
5         KODEX 200선물인버스2X   3315.0   75.0  -2.21%   3314.0  +38.93%   
..                     ...      ...    ...     ...      ...      ...   
545     KBSTAR 코스닥150선물인버스   5520.0  240.0  -4.17%   5523.0  +24.14%   
546        KBSTAR 200생활소비재   6605.0   10.0  -0.15%   6626.0   -9.22%   
547   KINDEX Fn K-뉴딜디지털플러스  10010.0  285.0  +2.93%  10018.0  -17.09%   
548               파워 중기국고채  96120.0   10.0  -0.01%  96202.0   -0.75%   
549     KBSTAR 팔라듐선물인버스(H)   4700.0   20.0  -0.42%   4772.0  +12.38%   

             거래량  거래대금(백만)  시가총액(억)  
1      3815046.0  118058.0  50710.0  
2      5475607.0   98250.0  40925.0  
3     32228858.0  463

종목명       현재가    전일비     등락률       NAV   3개월수익률  \
5      KODEX 200선물인버스2X    3315.0   75.0  -2.21%    3314.0  +38.93%   
3            KODEX 레버리지   14360.0  345.0  +2.46%   14394.0  -30.98%   
11           KODEX 단기채권  103395.0    0.0   0.00%  103396.0   +0.09%   
46    KODEX 코스닥150선물인버스    5265.0  215.0  -3.92%    5269.0  +24.40%   
22     KODEX 코스닥150레버리지    7865.0  555.0  +7.59%    7961.0  -40.69%   
..                  ...       ...    ...     ...       ...      ...   
420   마이다스 200커버드콜5%OTM   12260.0  205.0  +1.70%   12213.0  -13.55%   
423    KODEX MSCI Korea   14505.0  265.0  +1.86%   14516.0  -16.11%   
424       KBSTAR 모멘텀로우볼   13185.0   55.0  +0.42%   13207.0  -16.07%   
426         KBSTAR 주식혼합   35655.0  415.0  +1.18%   35630.0  -11.74%   
549  KBSTAR 팔라듐선물인버스(H)    4700.0   20.0  -0.42%    4772.0  +12.38%   

             거래량  거래대금(백만)  시가총액(억)  
5    196034897.0  648463.0  20059.0  
3     32228858.0  463304.0  22502.0  
11     2844435.0  294086.0  13295.0  
46    48377711.0  256636.0   3506.0  
22    26816282.0  208473.0   9068.0  
..           ...       ...      ...  
420          4.0       0.0     74.0  
423          9.0       0.0     73.0  
424          6.0       0.0     73.0  
426          4.0       0.0     71.0  
549        166.0       0.0     28.0  

[549 rows x 9 columns]

    총 549개 ETF 종목코드를 구해서 etfs 딕셔너리 생성 
    자동매매 프로그램에서 전체 ETF 종목에 대한 순회 처리때 활용

### 5.4.2 매수 목표가 계산 
   - 매수 목표가 = 금일 시작 가격 + (어제 최고가 - 어제 최저가)* K
   - K값은 일반적으로 0.5 이며, 높을수록 목표가 도달 가능성 낮아짐. 

In [19]:
def get_target_price(code):
    """매수 목표가를 반환한다."""
    try:
        time_now = datetime.now()
        str_today = time_now.strftime('%Y%m%d')
        ohlc = get_ohlc(code, 10)
        if str_today == str(ohlc.iloc[0].name):
            today_open = ohlc.iloc[0].open 
            lastday = ohlc.iloc[1]
        else:
            lastday = ohlc.iloc[0]                                      
            today_open = lastday[3]
        lastday_high = lastday[1]
        lastday_low = lastday[2]
        target_price = today_open + (lastday_high - lastday_low) * 0.5
        return target_price
    except Exception as ex:
        dbgout("`get_target_price() -> exception! " + str(ex) + "`")
        return None

In [20]:
get_target_price('A305080')

12085.0

### 5.4.3 이동평균값 조회
    PART4에서 배우 전략의 공통점: 시장 평가시, 한 가지 기준이 아닌 둘 이상의 기준으로 평가한다. 
    
    변동성 돌파 전략의 매수 조건은 현재가가 (시가 + 전일 변동폭의 K%)를 돌파할 때다. 
    여기에, 현재가가 5일 이동평균선과 10일 이동평균선 위에 있어야한다는 조건을 추가하고자 한다. 
    
    특정 종목의 종목코드와 이동평균의 기준일로 인수로 받아 해당 종목의 마지막 이동평균값을 반환한다. 

In [21]:
def get_movingaverage(code, window):
    """인자로 받은 종목에 대한 이동평균가격을 반환한다."""
    try:
        time_now = datetime.now()
        str_today = time_now.strftime('%Y%m%d')
        ohlc = get_ohlc(code, 20)
        if str_today == str(ohlc.iloc[0].name):
            lastday = ohlc.iloc[1].name
        else:
            lastday = ohlc.iloc[0].name
        closes = ohlc['close'].sort_index()         
        ma = closes.rolling(window=window).mean()
        return ma.loc[lastday]
    except Exception as ex:
        dbgout('get_movingavrg(' + str(window) + ') -> exception! ' + str(ex))
        return None 

In [23]:
get_movingaverage('A305080', 2)

12027.5

### 5.4.4 주식 매매 주문 방식 

#### 최유리 FOK 매수 주문

   - 오전 09:15 ~ 15:15까지는 일정 간격으로 매수 종목 후보군의 현재가, 매수 목표가, 5일 이동평균가, 10일 이동평균가를 구해서 매수 조건 부합 점검
   - 이 조건을 보고, 해당 종목 최유리 FOK 주문 방식으로 매수 
   - 이유: ETF 종목은 일반 주식 종목에 비해 거래량이 그리 많지 않음. 
           한꺼번에 원하는 수량만큼 매수시도하려면 가격이 비싸지는 경우 발생 가능. 
           이때, 최유리 FOK 매수 주문을 통해 현재 가장 낮은 매도호가에 매수 주문을 냄으로써 
           [원하는 가격에 원하는 수량만큼] 매수가 가능할 때만 계약이 체결되도록 함.
          

In [24]:
def buy_etf(code):
    """인자로 받은 종목을 최유리 지정가 FOK 조건으로 매수한다."""
    try:
        global bought_list      # 함수 내에서 값 변경을 하기 위해 global로 지정
        if code in bought_list: # 매수 완료 종목이면 더 이상 안 사도록 함수 종료
            #printlog('code:', code, 'in', bought_list)
            return False
        time_now = datetime.now()
        current_price, ask_price, bid_price = get_current_price(code) 
        target_price = get_target_price(code)    # 매수 목표가
        ma5_price = get_movingaverage(code, 5)   # 5일 이동평균가
        ma10_price = get_movingaverage(code, 10) # 10일 이동평균가
        buy_qty = 0        # 매수할 수량 초기화
        if ask_price > 0:  # 매도호가가 존재하면   
            buy_qty = buy_amount // ask_price  
        stock_name, stock_qty = get_stock_balance(code)  # 종목명과 보유수량 조회
        #printlog('bought_list:', bought_list, 'len(bought_list):',
        #    len(bought_list), 'target_buy_count:', target_buy_count)     
        if current_price > target_price and current_price > ma5_price \
            and current_price > ma10_price:  
            printlog(stock_name + '(' + str(code) + ') ' + str(buy_qty) +
                'EA : ' + str(current_price) + ' meets the buy condition!`')            
            cpTradeUtil.TradeInit()
            acc = cpTradeUtil.AccountNumber[0]      # 계좌번호
            accFlag = cpTradeUtil.GoodsList(acc, 1) # -1:전체,1:주식,2:선물/옵션                
            # 최유리 FOK 매수 주문 설정
            cpOrder.SetInputValue(0, "2")        # 2: 매수
            cpOrder.SetInputValue(1, acc)        # 계좌번호
            cpOrder.SetInputValue(2, accFlag[0]) # 상품구분 - 주식 상품 중 첫번째
            cpOrder.SetInputValue(3, code)       # 종목코드
            cpOrder.SetInputValue(4, buy_qty)    # 매수할 수량
            cpOrder.SetInputValue(7, "2")        # 주문조건 0:기본, 1:IOC, 2:FOK
            cpOrder.SetInputValue(8, "12")       # 주문호가 1:보통, 3:시장가
                                                 # 5:조건부, 12:최유리, 13:최우선 
            # 매수 주문 요청
            ret = cpOrder.BlockRequest() 
            printlog('최유리 FoK 매수 ->', stock_name, code, buy_qty, '->', ret)
            if ret == 4:
                remain_time = cpStatus.LimitRequestRemainTime
                printlog('주의: 연속 주문 제한에 걸림. 대기 시간:', remain_time/1000)
                time.sleep(remain_time/1000) 
                return False
            time.sleep(2)
            printlog('현금주문 가능금액 :', buy_amount)
            stock_name, bought_qty = get_stock_balance(code)
            printlog('get_stock_balance :', stock_name, stock_qty)
            if bought_qty > 0:
                bought_list.append(code)
                dbgout("`buy_etf("+ str(stock_name) + ' : ' + str(code) + 
                    ") -> " + str(bought_qty) + "EA bought!" + "`")
    except Exception as ex:
        dbgout("`buy_etf("+ str(code) + ") -> exception! " + str(ex) + "`")

#### 5.4.5 최유리 IOC 매도 주문 
    - 15:15 ~ 15:20까지 매수했던 종목을 모두 매도 청산
    - 장 마감 이후 해외 시장에 영향에 따른 변동성을 고려해 매수한 종목은 당일 매도을 진행한다. 

In [ ]:
def sell_all():
    """보유한 모든 종목을 최유리 지정가 IOC 조건으로 매도한다."""
    try:
        cpTradeUtil.TradeInit()
        acc = cpTradeUtil.AccountNumber[0]       # 계좌번호
        accFlag = cpTradeUtil.GoodsList(acc, 1)  # -1:전체, 1:주식, 2:선물/옵션   
        while True:    
            stocks = get_stock_balance('ALL') 
            total_qty = 0 
            for s in stocks:
                total_qty += s['qty'] 
            if total_qty == 0:
                return True
            for s in stocks:
                if s['qty'] != 0:                  
                    cpOrder.SetInputValue(0, "1")         # 1:매도, 2:매수
                    cpOrder.SetInputValue(1, acc)         # 계좌번호
                    cpOrder.SetInputValue(2, accFlag[0])  # 주식상품 중 첫번째
                    cpOrder.SetInputValue(3, s['code'])   # 종목코드
                    cpOrder.SetInputValue(4, s['qty'])    # 매도수량
                    cpOrder.SetInputValue(7, "1")   # 조건 0:기본, 1:IOC, 2:FOK
                    cpOrder.SetInputValue(8, "12")  # 호가 12:최유리, 13:최우선 
                    # 최유리 IOC 매도 주문 요청
                    ret = cpOrder.BlockRequest()
                    printlog('최유리 IOC 매도', s['code'], s['name'], s['qty'], 
                        '-> cpOrder.BlockRequest() -> returned', ret)
                    if ret == 4:
                        remain_time = cpStatus.LimitRequestRemainTime
                        printlog('주의: 연속 주문 제한, 대기시간:', remain_time/1000)
                time.sleep(1)
            time.sleep(30)
    except Exception as ex:
        dbgout("sell_all() -> exception! " + str(ex))
#메인 로직 코드
if __name__ == '__main__':                                                            # 1. EtFAlgoTrader.py 파일을 모듈로 실행할 때만 로직 동작하도록 처리
    try:
        symbol_list = ['A122630', 'A252670', 'A233740', 'A250780', 'A225130',
             'A280940', 'A261220', 'A217770', 'A295000', 'A176950']                   # 2. 매수할 ETF 후보 리스트.(내가 원하는걸로) 
        bought_list = []     # 매수 완료된 종목 리스트
        target_buy_count = 5 # 매수할 종목 수
        buy_percent = 0.19                                                            
        printlog('check_creon_system() :', check_creon_system())  # 크레온 접속 점검
        stocks = get_stock_balance('ALL')      # 보유한 모든 종목 조회
        total_cash = int(get_current_cash())   # 100% 증거금 주문 가능 금액 조회
        buy_amount = total_cash * buy_percent  # 종목별 주문 금액 계산
        printlog('100% 증거금 주문 가능 금액 :', total_cash)
        printlog('종목별 주문 비율 :', buy_percent)
        printlog('종목별 주문 금액 :', buy_amount)
        printlog('시작 시간 :', datetime.now().strftime('%m/%d %H:%M:%S'))
        soldout = False;

        while True:
            t_now = datetime.now()
            t_9 = t_now.replace(hour=9, minute=0, second=0, microsecond=0)
            t_start = t_now.replace(hour=9, minute=5, second=0, microsecond=0)
            t_sell = t_now.replace(hour=15, minute=15, second=0, microsecond=0)
            t_exit = t_now.replace(hour=15, minute=20, second=0,microsecond=0)
            today = datetime.today().weekday()
            if today == 5 or today == 6:                                            # 3. 토요일이나 일요일이면 자동 종료
                printlog('Today is', 'Saturday.' if today == 5 else 'Sunday.')
                sys.exit(0)
            if t_9 < t_now < t_start and soldout == False:
                soldout = True
                sell_all()
            if t_start < t_now < t_sell :                                          #  4. AM 09:05 ~ PM 03:15 : 매수
                for sym in symbol_list:
                    if len(bought_list) < target_buy_count:
                        buy_etf(sym)
                        time.sleep(1)
                if t_now.minute == 30 and 0 <= t_now.second <= 5: 
                    get_stock_balance('ALL')
                    time.sleep(5)
            if t_sell < t_now < t_exit:                                            # 5. PM 03:15 ~ PM 03:20 : 일괄 매도
                if sell_all() == True:
                    dbgout('`sell_all() returned True -> self-destructed!`')
                    sys.exit(0)
            if t_exit < t_now:                                                     # 6. PM 03:20 ~ :프로그램 종료
                dbgout('`self-destructed!`')
                sys.exit(0)
            time.sleep(3)
    except Exception as ex:
        dbgout('`main -> exception! ' + str(ex) + '`')

크레온 API 정상 접속 및 연동 확인되었습니다.
[07/05 13:50:24] check_creon_system() : None
[07/05 13:50:24] 계좌명: 변민찬
[07/05 13:50:24] 결제잔고수량 : 0
[07/05 13:50:24] 평가금액: 0
[07/05 13:50:24] 평가손익: 0
[07/05 13:50:24] 종목수: 0
[07/05 13:50:24] 100% 증거금 주문 가능 금액 : 0
[07/05 13:50:24] 종목별 주문 비율 : 0.19
[07/05 13:50:24] 종목별 주문 금액 : 0.0
[07/05 13:50:24] 시작 시간 : 07/05 13:50:24
[07/05 13:50:33] KBSTAR 국채선물10년(A295000) 0.0EA : 50245 meets the buy condition!`
[07/05 13:50:33] 최유리 FoK 매수 -> KBSTAR 국채선물10년 A295000 0.0 -> 5
[07/05 13:50:35] 현금주문 가능금액 : 0.0
[07/05 13:50:35] get_stock_balance : KBSTAR 국채선물10년 0
[07/05 13:50:50] KBSTAR 국채선물10년(A295000) 0.0EA : 50245 meets the buy condition!`
[07/05 13:50:50] 최유리 FoK 매수 -> KBSTAR 국채선물10년 A295000 0.0 -> 5
[07/05 13:50:52] 현금주문 가능금액 : 0.0
[07/05 13:50:52] get_stock_balance : KBSTAR 국채선물10년 0
[07/05 13:51:07] KBSTAR 국채선물10년(A295000) 0.0EA : 50245 meets the buy condition!`
[07/05 13:51:07] 최유리 FoK 매수 -> KBSTAR 국채선물10년 A295000 0.0 -> 5
[07/05 13:51:09] 현금주문 가능금액 : 0.0
[07/05 13:51:0